iroh-gateway/src/core.rs

```diff
@@ -0,0 +1,412 @@
+        .await;
+    let body = match body {
+        Ok(b) => b,
+        Err(e) => {
+            let msg = format!("{}", e);
+            return error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+        }
+    };
+
+    set_content_disposition_headers(
+        headers,
+        format!("{}.bin", req.cid).as_str(),
+        DISPOSITION_ATTACHMENT,
+    );
+    response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+}
+
+async fn serve_car(
+    req: &Request,
+    client: Client,
+    headers: &mut HashMap<String, String>,
+) -> Result<GatewayResponse, GatewayError> {
+    let body = client
+        .get_file(format!("{}", req.full_content_path).as_str())
+        .await;
+    let body = match body {
+        Ok(b) => b,
+        Err(e) => {
+            let msg = format!("{}", e);
+            return error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+        }
+    };
+    set_content_disposition_headers(
+        headers,
+        format!("{}.car", req.cid).as_str(),
+        DISPOSITION_ATTACHMENT,
+    );
+    response(StatusCode::OK, body::boxed(Body::from(body)), &headers)
+}
+
+async fn serve_html(
+    req: &Request,
+    client: Client,
+    headers: &mut HashMap<String, String>,
+) -> Result<GatewayResponse, GatewayError> {
+    let body = client
+        .get_file(format!("{}", req.full_content_path).as_str())
+        .await;
+    let body = match body {
+        Ok(b) => b,
+        Err(e) => {
+            let msg = format!("{}", e);
+            return error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+        }
+    };
+    response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+}
+
+async fn serve_fs(
+    req: &Request,
+    client: Client,
+    headers: &mut HashMap<String, String>,
+) -> Result<GatewayResponse, GatewayError> {
+    let body = client
+        .get_file(format!("{}", req.full_content_path).as_str())
+        .await;
+    let body = match body {
+        Ok(b) => b,
+        Err(e) => {
+            let msg = format!("{}", e);
+            return error(StatusCode::INTERNAL_SERVER_ERROR, &msg);
+        }
+    };
+    let name = add_content_disposition_headers(
+        headers,
+        &req.query_file_name,
+        &req.content_path,
+        req.download,
+    );
+    add_content_type_headers(headers, &name);
+    response(StatusCode::OK, body::boxed(Body::from(body)), headers)
+}
+
+fn add_user_headers(headers: &mut HashMap<String, String>, user_headers: HashMap<String, String>) {
+    for (key, value) in user_headers {
+        headers.insert(key, value);
+    }
+}
+
+fn add_content_type_headers(headers: &mut HashMap<String, String>, name: &str) {
+    let guess = mime_guess::from_path(name);
+    let content_type = guess.first_or_octet_stream().to_string();
+    headers.insert(CONTENT_TYPE.to_string(), content_type);
+}
+
+fn add_content_disposition_headers(
+    headers: &mut HashMap<String, String>,
+    filename: &str,
+    content_path: &str,
+    download: bool,
```

Comment:
this is hard to understand naming wise, so `bool` flags usually should carry a prefix, sth like `should_download` or similar